File: web/src/config/homeContent.ts
import type { HomeContent, LobbySetting } from '@/types/setting'

/**
 * 主页文案（home-intro + latest posts）的字段目录。
 *
 * 这一组字段聚合在 `site_settings.home_content` 一列 JSON 中，与
 * `theme_elements` 的差异在于：不分主题，整组替换。后台 Settings 页的
 * 「主页文案」段直接遍历 `homeContentFields` 渲染；新增字段只动本文件
 * + 后端 `homecontent` 包的常量。
 */
export type HomeContentKey = keyof HomeContent

export interface HomeContentField {
  key: HomeContentKey
  label: string
  hint: string
  max: number
  multiline?: boolean
}

// 与后端 homecontent.go 的 *MaxRunes 常量保持一一对应。
export const homeContentFields: ReadonlyArray<HomeContentField> = [
  {
    key: 'home_profile_kicker',
    label: '主页 Profile 小标',
    hint: '头像右侧作者身份上方的英文小标；最多 32 字。',
    max: 32,
  },
  {
    key: 'home_heading_prefix',
    label: '主页问候语前缀',
    hint: '作者名前的称呼语，例如 "你好，我是"；最多 16 字。',
    max: 16,
  },
  {
    key: 'home_status_fallback',
    label: '主页状态回落文案',
    hint: '当没有启用中的 Site Notice 时显示；最多 80 字。',
    max: 80,
  },
  {
    key: 'home_intro_heading',
    label: '主页介绍标题',
    hint: 'home-intro 卡片顶部的大标题；最多 80 字。',
    max: 80,
  },
  {
    key: 'home_intro_paragraph',
    label: '主页介绍段落',
    hint: 'home-intro 卡片标题下的多行描述；最多 240 字。',
    max: 240,
    multiline: true,
  },
  {
    key: 'home_action_view_recent_label',
    label: '"查看最近发布"按钮',
    hint: 'home-intro 卡片左下主操作按钮；最多 16 字。',
    max: 16,
  },
  {
    key: 'home_action_view_archive_label',
    label: '"浏览全部归档"按钮',
    hint: 'home-intro 卡片右下副操作按钮；最多 16 字。',
    max: 16,
  },
  {
    key: 'home_latest_kicker',
    label: '最近发布小标',
    hint: '"Latest posts" 段上方小标；最多 32 字。',
    max: 32,
  },
  {
    key: 'home_latest_heading',
    label: '最近发布标题',
    hint: '"Latest posts" 段标题；最多 64 字。',
    max: 64,
  },
  {
    key: 'home_latest_view_all_label',
    label: '"查看全部归档"链接',
    hint: '"Latest posts" 段右上角"查看全部"链接文案；最多 16 字。',
    max: 16,
  },
  {
    key: 'home_latest_empty_title',
    label: '最近发布空状态标题',
    hint: '没有任何已发布文章时显示的标题；最多 64 字。',
    max: 64,
  },
  {
    key: 'home_topics_kicker',
    label: '专题目录小标',
    hint: '"Topics" 段上方小标；最多 32 字。',
    max: 32,
  },
  {
    key: 'home_topics_heading',
    label: '专题目录标题',
    hint: '"Topics" 段标题；最多 64 字。',
    max: 64,
  },
  {
    key: 'home_notice_kicker',
    label: '站点公告小标',
    hint: '启用中的 Site Notice 卡片上方小标；最多 32 字。',
    max: 32,
  },
  {
    key: 'home_notice_action_label',
    label: '"继续阅读"按钮',
    hint: 'Site Notice 卡片上的跳转按钮；最多 16 字。',
    max: 16,
  },
]

export const defaultHomeContent: Readonly<HomeContent> = {
  home_profile_kicker: 'Blog keeper · Solitude',
  home_heading_prefix: '你好，我是',
  home_status_fallback: '持续记录技术、设计与生活',
  home_intro_heading: '一份持续更新的博客，也是公开的思考现场',
  home_intro_paragraph:
    '在这里记录工程实践、设计系统与构建过程。文章保留可复用的方法，也保留问题发生时的真实判断。',
  home_action_view_recent_label: '查看最近发布',
  home_action_view_archive_label: '浏览全部归档',
  home_latest_kicker: 'Latest posts',
  home_latest_heading: '最近发布的博客',
  home_latest_view_all_label: '查看全部归档',
  home_latest_empty_title: '暂时还没有发布文章',
  home_topics_kicker: 'Topics',
  home_topics_heading: '从这些专题进入',
  home_notice_kicker: 'Site notice',
  home_notice_action_label: '继续阅读',
}

export function createDefaultHomeContent(): HomeContent {
  return { ...defaultHomeContent }
}

type HomeContentInput = Partial<HomeContent> | null | undefined

function resolveText(value: unknown, fallback: string) {
  return typeof value === 'string' && value.trim() ? value.trim() : fallback
}

export function normalizeHomeContent(value?: HomeContentInput): HomeContent {
  return {
    home_profile_kicker: resolveText(value?.home_profile_kicker, defaultHomeContent.home_profile_kicker),
    home_heading_prefix: resolveText(value?.home_heading_prefix, defaultHomeContent.home_heading_prefix),
    home_status_fallback: resolveText(value?.home_status_fallback, defaultHomeContent.home_status_fallback),
    home_intro_heading: resolveText(value?.home_intro_heading, defaultHomeContent.home_intro_heading),
    home_intro_paragraph: resolveText(
      value?.home_intro_paragraph,
      defaultHomeContent.home_intro_paragraph,
    ),
    home_action_view_recent_label: resolveText(
      value?.home_action_view_recent_label,
      defaultHomeContent.home_action_view_recent_label,
    ),
    home_action_view_archive_label: resolveText(
      value?.home_action_view_archive_label,
      defaultHomeContent.home_action_view_archive_label,
    ),
    home_latest_kicker: resolveText(value?.home_latest_kicker, defaultHomeContent.home_latest_kicker),
    home_latest_heading: resolveText(value?.home_latest_heading, defaultHomeContent.home_latest_heading),
    home_latest_view_all_label: resolveText(
      value?.home_latest_view_all_label,
      defaultHomeContent.home_latest_view_all_label,
    ),
    home_latest_empty_title: resolveText(
      value?.home_latest_empty_title,
      defaultHomeContent.home_latest_empty_title,
    ),
    home_topics_kicker: resolveText(value?.home_topics_kicker, defaultHomeContent.home_topics_kicker),
    home_topics_heading: resolveText(value?.home_topics_heading, defaultHomeContent.home_topics_heading),
    home_notice_kicker: resolveText(value?.home_notice_kicker, defaultHomeContent.home_notice_kicker),
    home_notice_action_label: resolveText(
      value?.home_notice_action_label,
      defaultHomeContent.home_notice_action_label,
    ),
  }
}

/**
 * 渲染层统一入口：始终返回一个完整的 HomeContent，未提供或字段为空白时
 * 回退到 defaultHomeContent。这样模板中可以直接 `{{ homeContent.xxx }}`
 * 而不必处理 undefined / 空字符串。
 */
export function resolveLobbyHomeContent(setting: LobbySetting | null | undefined): HomeContent {
  if (!setting) {
    return { ...defaultHomeContent }
  }
  return normalizeHomeContent(setting.home_content)
}

File: server/internal/homecontent/homecontent.go
// Package homecontent owns the per-site home page copy contract. Values are
// theme-independent so they live as a single JSON column on site_settings
// instead of inside appearance.ThemeElements.
package homecontent

import (
	"strings"
	"unicode/utf8"
)

const (
	// HomeProfileKickerMaxRunes caps the small label rendered above the
	// author identity block on the home page (e.g. "Blog keeper · Solitude").
	HomeProfileKickerMaxRunes = 32
	// HomeHeadingPrefixMaxRunes caps the short greeting prefix rendered
	// before the author name (e.g. "你好，我是").
	HomeHeadingPrefixMaxRunes = 16
	// HomeStatusFallbackMaxRunes caps the status line shown when there is
	// no active site notice.
	HomeStatusFallbackMaxRunes = 80
	// HomeIntroHeadingMaxRunes caps the heading rendered inside the
	// home-intro aside.
	HomeIntroHeadingMaxRunes = 80
	// HomeIntroParagraphMaxRunes caps the multi-line description rendered
	// below the home-intro heading.
	HomeIntroParagraphMaxRunes = 240
	// HomeActionLabelMaxRunes caps the two short action button labels on
	// the home-intro aside.
	HomeActionLabelMaxRunes = 16
	// HomeLatestKickerMaxRunes caps the small label rendered above the
	// "Latest posts" heading.
	HomeLatestKickerMaxRunes = 32
	// HomeLatestHeadingMaxRunes caps the heading rendered above the latest
	// articles list.
	HomeLatestHeadingMaxRunes = 64
	// HomeLatestViewAllLabelMaxRunes caps the "view all" link label.
	HomeLatestViewAllLabelMaxRunes = 16
	// HomeLatestEmptyTitleMaxRunes caps the title shown when there are no
	// published articles.
	HomeLatestEmptyTitleMaxRunes = 64
	// HomeTopicsKickerMaxRunes caps the small label rendered above the
	// topics directory heading.
	HomeTopicsKickerMaxRunes = 32
	// HomeTopicsHeadingMaxRunes caps the heading rendered above the topics
	// directory links.
	HomeTopicsHeadingMaxRunes = 64
	// HomeNoticeKickerMaxRunes caps the small label rendered above the site
	// notice heading.
	HomeNoticeKickerMaxRunes = 32
	// HomeNoticeActionLabelMaxRunes caps the short action link label on the
	// site notice card.
	HomeNoticeActionLabelMaxRunes = 16
)

// HomeContent is the full set of theme-independent home page copy fields.
// Every field is required so the admin form and the public renderer can
// rely on a complete object.
type HomeContent struct {
	HomeProfileKicker          string `json:"home_profile_kicker"`
	HomeHeadingPrefix          string `json:"home_heading_prefix"`
	HomeStatusFallback         string `json:"home_status_fallback"`
	HomeIntroHeading           string `json:"home_intro_heading"`
	HomeIntroParagraph         string `json:"home_intro_paragraph"`
	HomeActionViewRecentLabel  string `json:"home_action_view_recent_label"`
	HomeActionViewArchiveLabel string `json:"home_action_view_archive_label"`
	HomeLatestKicker           string `json:"home_latest_kicker"`
	HomeLatestHeading          string `json:"home_latest_heading"`
	HomeLatestViewAllLabel     string `json:"home_latest_view_all_label"`
	HomeLatestEmptyTitle       string `json:"home_latest_empty_title"`
	HomeTopicsKicker           string `json:"home_topics_kicker"`
	HomeTopicsHeading          string `json:"home_topics_heading"`
	HomeNoticeKicker           string `json:"home_notice_kicker"`
	HomeNoticeActionLabel      string `json:"home_notice_action_label"`
}

// DefaultHomeContent returns a complete, independently allocated copy of the
// shipped home page copy. The values match the strings that used to be
// hard-coded in HomePage.vue so the migration is byte-for-byte identical
// when no admin override is set.
func DefaultHomeContent() HomeContent {
	return HomeContent{
		HomeProfileKicker:          "Blog keeper · Solitude",
		HomeHeadingPrefix:          "你好，我是",
		HomeStatusFallback:         "持续记录技术、设计与生活",
		HomeIntroHeading:           "一份持续更新的博客，也是公开的思考现场",
		HomeIntroParagraph:         "在这里记录工程实践、设计系统与构建过程。文章保留可复用的方法，也保留问题发生时的真实判断。",
		HomeActionViewRecentLabel:  "查看最近发布",
		HomeActionViewArchiveLabel: "浏览全部归档",
		HomeLatestKicker:           "Latest posts",
		HomeLatestHeading:          "最近发布的博客",
		HomeLatestViewAllLabel:     "查看全部归档",
		HomeLatestEmptyTitle:       "暂时还没有发布文章",
		HomeTopicsKicker:           "Topics",
		HomeTopicsHeading:          "从这些专题进入",
		HomeNoticeKicker:           "Site notice",
		HomeNoticeActionLabel:      "继续阅读",
	}
}

// IsValidHomeContent reports whether every field satisfies its rune cap.
// Empty values are valid so NormalizeHomeContent can reset them to the
// shipped defaults.
func IsValidHomeContent(value HomeContent) bool {
	return !exceedsRuneLimit(value.HomeProfileKicker, HomeProfileKickerMaxRunes) &&
		!exceedsRuneLimit(value.HomeHeadingPrefix, HomeHeadingPrefixMaxRunes) &&
		!exceedsRuneLimit(value.HomeStatusFallback, HomeStatusFallbackMaxRunes) &&
		!exceedsRuneLimit(value.HomeIntroHeading, HomeIntroHeadingMaxRunes) &&
		!exceedsRuneLimit(value.HomeIntroParagraph, HomeIntroParagraphMaxRunes) &&
		!exceedsRuneLimit(value.HomeActionViewRecentLabel, HomeActionLabelMaxRunes) &&
		!exceedsRuneLimit(value.HomeActionViewArchiveLabel, HomeActionLabelMaxRunes) &&
		!exceedsRuneLimit(value.HomeLatestKicker, HomeLatestKickerMaxRunes) &&
		!exceedsRuneLimit(value.HomeLatestHeading, HomeLatestHeadingMaxRunes) &&
		!exceedsRuneLimit(value.HomeLatestViewAllLabel, HomeLatestViewAllLabelMaxRunes) &&
		!exceedsRuneLimit(value.HomeLatestEmptyTitle, HomeLatestEmptyTitleMaxRunes) &&
		!exceedsRuneLimit(value.HomeTopicsKicker, HomeTopicsKickerMaxRunes) &&
		!exceedsRuneLimit(value.HomeTopicsHeading, HomeTopicsHeadingMaxRunes) &&
		!exceedsRuneLimit(value.HomeNoticeKicker, HomeNoticeKickerMaxRunes) &&
		!exceedsRuneLimit(value.HomeNoticeActionLabel, HomeNoticeActionLabelMaxRunes)
}

// NormalizeHomeContent returns a complete HomeContent by trimming every
// field and falling back to the corresponding default when the input is
// empty or longer than the per-field rune cap.
func NormalizeHomeContent(value HomeContent) HomeContent {
	defaults := DefaultHomeContent()
	return HomeContent{
		HomeProfileKicker:          normalizeHomeText(value.HomeProfileKicker, defaults.HomeProfileKicker, HomeProfileKickerMaxRunes),
		HomeHeadingPrefix:          normalizeHomeText(value.HomeHeadingPrefix, defaults.HomeHeadingPrefix, HomeHeadingPrefixMaxRunes),
		HomeStatusFallback:         normalizeHomeText(value.HomeStatusFallback, defaults.HomeStatusFallback, HomeStatusFallbackMaxRunes),
		HomeIntroHeading:           normalizeHomeText(value.HomeIntroHeading, defaults.HomeIntroHeading, HomeIntroHeadingMaxRunes),
		HomeIntroParagraph:         normalizeHomeText(value.HomeIntroParagraph, defaults.HomeIntroParagraph, HomeIntroParagraphMaxRunes),
		HomeActionViewRecentLabel:  normalizeHomeText(value.HomeActionViewRecentLabel, defaults.HomeActionViewRecentLabel, HomeActionLabelMaxRunes),
		HomeActionViewArchiveLabel: normalizeHomeText(value.HomeActionViewArchiveLabel, defaults.HomeActionViewArchiveLabel, HomeActionLabelMaxRunes),
		HomeLatestKicker:           normalizeHomeText(value.HomeLatestKicker, defaults.HomeLatestKicker, HomeLatestKickerMaxRunes),
		HomeLatestHeading:          normalizeHomeText(value.HomeLatestHeading, defaults.HomeLatestHeading, HomeLatestHeadingMaxRunes),
		HomeLatestViewAllLabel:     normalizeHomeText(value.HomeLatestViewAllLabel, defaults.HomeLatestViewAllLabel, HomeLatestViewAllLabelMaxRunes),
		HomeLatestEmptyTitle:       normalizeHomeText(value.HomeLatestEmptyTitle, defaults.HomeLatestEmptyTitle, HomeLatestEmptyTitleMaxRunes),
		HomeTopicsKicker:           normalizeHomeText(value.HomeTopicsKicker, defaults.HomeTopicsKicker, HomeTopicsKickerMaxRunes),
		HomeTopicsHeading:          normalizeHomeText(value.HomeTopicsHeading, defaults.HomeTopicsHeading, HomeTopicsHeadingMaxRunes),
		HomeNoticeKicker:           normalizeHomeText(value.HomeNoticeKicker, defaults.HomeNoticeKicker, HomeNoticeKickerMaxRunes),
		HomeNoticeActionLabel:      normalizeHomeText(value.HomeNoticeActionLabel, defaults.HomeNoticeActionLabel, HomeNoticeActionLabelMaxRunes),
	}
}

// CloneHomeContent returns a value-level copy. All fields are strings so a
// shallow struct copy is sufficient.
func CloneHomeContent(value HomeContent) HomeContent {
	return value
}

func exceedsRuneLimit(value string, limit int) bool {
	return utf8.RuneCountInString(strings.TrimSpace(value)) > limit
}

func normalizeHomeText(value string, fallback string, limit int) string {
	value = strings.TrimSpace(value)
	if value == "" || utf8.RuneCountInString(value) > limit {
		return fallback
	}
	return value
}
